### DISCUSSION TITLE
unique_ptr使用问题

### DISCUSSION THREAD

**Discussion by cellphonef** [chenshuo/muduo#691] [General]
如TcpServer中的acceptor_成员和TcpConneciton中的socket_以及channel_成员，都使用unique_ptr来管理，个人觉得可以直接使用对象即可，想问muduo使用unique_ptr来管理是出于什么原因？减少头文件的依赖吗？

**Comment by xb1520** [comment]
他们之间相互引用了,所以只能使用指针不能使用实体

**Reply by cellphonef** [reply]
socket和channel相互引用？我看并没有啊 只是都包含fd的数据成员？可以举个例子吗 谢谢

**Comment by xb1520** [comment]
**TcpServer**中的**acceptor_**成员后面有一句注释**avoid revealing Acceptor**，翻译成中文就是避免暴露接受者，就是不让类**Acceptor**暴漏在当前头文件，所以只能用指针，为了自动释放，就用了**unique_ptr**，**TcpConneciton**中的**socket_**以及**channel_**成员也是同理，上面也有一行注释**we don't expose those classes to client.**，我们不向客户端公开这些类。不好意思我最开始说的互相引用指的是**Channel**和**EventLoop**，我以为你在说这个，刚刚去仔细看了下你的问题和源码

**Comment by chenshuo** [comment]
本意确实是“减少头文件的依赖”。
